uncover/session.go

```diff
@@ -36,16 +40,60 @@ func NewSession(keys *Keys, retryMax, timeout int) (*Session, error) {
 	options.RetryWaitMax = time.Duration(timeout) * time.Second
 	client := retryablehttp.NewWithHTTPClient(httpclient, options)
 
+	var defaultRateLimits = map[string]*ratelimit.Options{
+		"shodan":     {Key: "shodan", MaxCount: 1, Duration: time.Second, IsUnlimited: true},
+		"shodan-idb": {Key: "shodan-idb", MaxCount: 1, Duration: time.Second, IsUnlimited: true},
+		"fofa":       {Key: "fofa", MaxCount: 1, Duration: time.Second, IsUnlimited: true},
+		"censys":     {Key: "censys", MaxCount: 1, Duration: 3 * time.Second, IsUnlimited: false},
+		"quake":      {Key: "quake", MaxCount: 1, Duration: time.Second, IsUnlimited: true},
+		"hunter":     {Key: "hunter", MaxCount: 15, Duration: time.Second, IsUnlimited: false},
+		"zoomeye":    {Key: "zoomeye", MaxCount: 1, Duration: time.Second, IsUnlimited: false},
+		"netlas":     {Key: "netlas", MaxCount: 1, Duration: time.Second, IsUnlimited: false},
+		"criminalip": {Key: "criminalip", MaxCount: 1, Duration: time.Second, IsUnlimited: true},
+		"publicwww":  {Key: "publicwww", MaxCount: 1, Duration: time.Minute, IsUnlimited: false},
+	}
+
 	session := &Session{
-		Client:   client,
-		Keys:     keys,
-		RetryMax: retryMax,
+		Client:     client,
+		Keys:       keys,
+		RetryMax:   retryMax,
+		RateLimits: &ratelimit.MultiLimiter{},
+	}
+
+	for i, engine := range engines {
+		var err error
+		rateLimitOpts := defaultRateLimits[engine]
+		if rateLimitOpts == nil {
+			return nil, fmt.Errorf("no default rate limit found for engine %s", engine)
+		}
+		if i == 0 {
+			session.RateLimits, err = ratelimit.NewMultiLimiter(context.Background(), defaultRateLimits[engine])
+			if err != nil {
+				return &Session{}, err
+			}
+		} else {
+			if rateLimit > 0 {
+				rateLimitOpts.MaxCount = uint(rateLimit)
+				rateLimitOpts.Duration = duration
+			}
+			rateLimitOpts.IsUnlimited = rateLimit == 0
+			err = session.RateLimits.Add(rateLimitOpts)
+			if err != nil {
+				return nil, err
+			}
+		}
 	}
 
 	return session, nil
 }
 
-func (s *Session) Do(request *retryablehttp.Request) (*http.Response, error) {
+func (s *Session) Do(request *retryablehttp.Request, sources ...string) (*http.Response, error) {
```

Comment:
As we make one request only, shouldn't this accept only one source to identify the specific rate limit?

Reply:
yeah, you are right, we can directly use an engine and pass the `agent.Name()`